[Bug] [SQL Transform] Cast bigint to timestamp error

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/seatunnel/issues?q=is%3Aissue+label%3A%22bug%22) and found no similar issues.


### What happened

2023-11-08 14:39:35,068 ERROR org.apache.seatunnel.core.starter.SeaTunnel - Exception StackTrace:org.apache.seatunnel.core.starter.exception.CommandExecuteException: SeaTunnel job executed failed
	at org.apache.seatunnel.core.starter.seatunnel.command.ClientExecuteCommand.execute(ClientExecuteCommand.java:192)
	at org.apache.seatunnel.core.starter.SeaTunnel.run(SeaTunnel.java:40)
	at org.apache.seatunnel.example.engine.TransformV2Example.main(TransformV2Example.java:43)
Caused by: org.apache.seatunnel.engine.common.exception.SeaTunnelEngineException: java.lang.IndexOutOfBoundsException: Index: 3, Size: 2
	at java.util.ArrayList.rangeCheck(ArrayList.java:657)
	at java.util.ArrayList.get(ArrayList.java:433)
	at org.apache.seatunnel.transform.sql.zeta.functions.SystemFunction.castAs(SystemFunction.java:118)
	at org.apache.seatunnel.transform.sql.zeta.ZetaSQLFunction.executeCastExpr(ZetaSQLFunction.java:448)
	at org.apache.seatunnel.transform.sql.zeta.ZetaSQLFunction.computeForValue(ZetaSQLFunction.java:230)
	at org.apache.seatunnel.transform.sql.zeta.ZetaSQLEngine.project(ZetaSQLEngine.java:261)
	at org.apache.seatunnel.transform.sql.zeta.ZetaSQLEngine.transformBySQL(ZetaSQLEngine.java:228)
	at org.apache.seatunnel.transform.sql.SQLTransform.transformRow(SQLTransform.java:115)
	at org.apache.seatunnel.transform.common.AbstractSeaTunnelTransform.map(AbstractSeaTunnelTransform.java:40)
	at org.apache.seatunnel.transform.common.AbstractSeaTunnelTransform.map(AbstractSeaTunnelTransform.java:26)
	at org.apache.seatunnel.engine.server.task.flow.TransformFlowLifeCycle.received(TransformFlowLifeCycle.java:94)
	at org.apache.seatunnel.engine.server.task.flow.TransformFlowLifeCycle.received(TransformFlowLifeCycle.java:37)
	at org.apache.seatunnel.engine.server.task.SeaTunnelSourceCollector.sendRecordToNext(SeaTunnelSourceCollector.java:208)
	at org.apache.seatunnel.engine.server.task.SeaTunnelSourceCollector.collect(SeaTunnelSourceCollector.java:123)
	at org.apache.seatunnel.connectors.seatunnel.fake.source.FakeDataGenerator.collectFakedRows(FakeDataGenerator.java:99)
	at org.apache.seatunnel.connectors.seatunnel.fake.source.FakeSourceReader.pollNext(FakeSourceReader.java:75)
	at org.apache.seatunnel.engine.server.task.flow.SourceFlowLifeCycle.collect(SourceFlowLifeCycle.java:150)
	at org.apache.seatunnel.engine.server.task.SourceSeaTunnelTask.collect(SourceSeaTunnelTask.java:117)
	at org.apache.seatunnel.engine.server.task.SeaTunnelTask.stateProcess(SeaTunnelTask.java:168)
	at org.apache.seatunnel.engine.server.task.SourceSeaTunnelTask.call(SourceSeaTunnelTask.java:122)
	at org.apache.seatunnel.engine.server.TaskExecutionService$BlockingWorker.run(TaskExecutionService.java:634)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
	at java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

	at org.apache.seatunnel.engine.client.job.ClientJobProxy.waitForJobComplete(ClientJobProxy.java:122)
	at org.apache.seatunnel.core.starter.seatunnel.command.ClientExecuteCommand.execute(ClientExecuteCommand.java:185)
	... 2 more

### SeaTunnel Version

all

### SeaTunnel Config

```conf
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
######
###### This config file is a demonstration of streaming processing in seatunnel config
######

env {
  execution.parallelism = 1
  job.mode = "BATCH"
  checkpoint.interval = 10000
}

source {
  FakeSource {
    result_table_name = "fake"
    schema = {
      fields {
        id = "bigint"
        name = "string"
        c1 = "string"
        c2 = "timestamp"
        c3 = "string"
        c4 = "bigint"
      }
    }
    rows = [
      {fields = [1, "Joy Ding", "12.4", "2012-12-21T12:34:56", null, 100000000000], kind = INSERT}
    ]
  }
}

transform {
  Sql {
    source_table_name = "fake"
    result_table_name = "fake1"
    query = "select cast(id as STRING) as id, cast(id as INT) as id2, cast(id as DOUBLE) as id3 , cast(c1 as double) as c1_1, cast(c1 as DECIMAL(10,2)) as c1_2, cast(c2 as DATE) as c2_1, coalesce(c3,'Unknown') c3_1, ifnull(c3,'Unknown') c3_2, ifnull(nullif(name,'Joy Ding'),'NULL') name1, nullif(name,'Joy Ding_') name2, cast(c4 as timestamp) as c4_1, cast(c4 as decimal(16,4)) as c4_2 from fake"
  }
}

sink {
  Console {
    source_table_name = "fake1"
  }
  Assert {
    source_table_name = "fake1"
    rules = {
      field_rules = [
        {
          field_name = "id"
          field_type = "string"
          field_value = [
            {equals_to = "1"}
          ]
        },
        {
          field_name = "id2"
          field_type = "int"
          field_value = [
            {equals_to = 1}
          ]
        },
        {
          field_name = "id3"
          field_type = "double"
          field_value = [
            {equals_to = 1}
          ]
        },
        {
          field_name = "c1_1"
          field_type = "double"
          field_value = [
            {equals_to = 12.4}
          ]
        },
        {
          field_name = "c1_2"
          field_type = "decimal(10,2)"
          field_value = [
            {equals_to = "12.40"}
          ]
        },
        {
          field_name = "c2_1"
          field_type = "date"
          field_value = [
            {equals_to = "2012-12-21"}
          ]
        },
        {
          field_name = "c3_1"
          field_type = "string"
          field_value = [
            {equals_to = "Unknown"}
          ]
        },
        {
          field_name = "c3_2"
          field_type = "string"
          field_value = [
            {equals_to = "Unknown"}
          ]
        },
        {
          field_name = "name1"
          field_type = "string"
          field_value = [
            {equals_to = "NULL"}
          ]
        },
        {
          field_name = "name2"
          field_type = "string"
          field_value = [
            {equals_to = "Joy Ding"}
          ]
        },
        {
          field_name = "c4_1"
          field_type = "timestamp"
          field_value = [
            {equals_to = 100000000000}
          ]
        },
        {
          field_name = "c4_2"
          field_type = "decimal(16,4)"
          field_value = [
            {equals_to = "100000000000.0000"}
          ]
        }
      ]
    }
  }
}
```


### Running Command

```shell
*
```


### Error Exception

```log
*
```


### Zeta or Flink or Spark Version

_No response_

### Java or Scala Version

_No response_

### Screenshots

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/seatunnel-connectors-v2/connector-assert/src/main/java/org/apache/seatunnel/connectors/seatunnel/assertion/rule/AssertRuleParser.java
+++ b/seatunnel-connectors-v2/connector-assert/src/main/java/org/apache/seatunnel/connectors/seatunnel/assertion/rule/AssertRuleParser.java
@@ -91,6 +91,12 @@ public class AssertRuleParser {
     }
 
     private SeaTunnelDataType<?> getFieldType(String fieldTypeStr) {
+        if (fieldTypeStr.toLowerCase().startsWith("decimal(")) {
+            String lengthAndScale =
+                    fieldTypeStr.toLowerCase().replace("decimal(", "").replace(")", "");
+            String[] split = lengthAndScale.split(",");
+            return new DecimalType(Integer.valueOf(split[0]), Integer.valueOf(split[1]));
+        }
         return TYPES.get(fieldTypeStr.toLowerCase());
     }
 
@@ -110,6 +116,5 @@ public class AssertRuleParser {
         TYPES.put("datetime", LocalTimeType.LOCAL_DATE_TIME_TYPE);
         TYPES.put("date", LocalTimeType.LOCAL_DATE_TYPE);
         TYPES.put("time", LocalTimeType.LOCAL_TIME_TYPE);
-        TYPES.put("decimal", new DecimalType(38, 18));
     }
 }

--- a/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-3/src/test/resources/jdbc_hive_source_and_assert.conf
+++ b/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-3/src/test/resources/jdbc_hive_source_and_assert.conf
@@ -127,12 +127,12 @@ sink{
           },
         {
           field_name = hive_e2e_source_table.decimal_column
-          field_type = decimal
+          field_type = "decimal(10,2)"
           field_value = [{equals_to = 42.12}]
           },
         {
          field_name = hive_e2e_source_table.numeric_column
-         field_type = decimal
+         field_type = "decimal(10,2)"
          field_value = [{equals_to = 42.12}]
          },
         ]

--- a/seatunnel-e2e/seatunnel-transforms-v2-e2e/seatunnel-transforms-v2-e2e-part-2/src/test/resources/json_path_transform/json_path_basic_type_test.conf
+++ b/seatunnel-e2e/seatunnel-transforms-v2-e2e/seatunnel-transforms-v2-e2e-part-2/src/test/resources/json_path_transform/json_path_basic_type_test.conf
@@ -159,7 +159,7 @@ sink {
           },
           {
             field_name = c1_decimal
-            field_type = decimal
+            field_type = "decimal(4,2)"
             field_value = [
               {
                 rule_type = NOT_NULL

--- a/seatunnel-e2e/seatunnel-transforms-v2-e2e/seatunnel-transforms-v2-e2e-part-2/src/test/resources/sql_transform/func_system.conf
+++ b/seatunnel-e2e/seatunnel-transforms-v2-e2e/seatunnel-transforms-v2-e2e-part-2/src/test/resources/sql_transform/func_system.conf
@@ -34,10 +34,13 @@ source {
         c1 = "string"
         c2 = "timestamp"
         c3 = "string"
+        c4 = "bigint"
+        c5 = "int"
+        c6 = "int"
       }
     }
     rows = [
-      {fields = [1, "Joy Ding", "12.4", "2012-12-21T12:34:56", null], kind = INSERT}
+      {fields = [1, "Joy Ding", "12.4", "2012-12-21T12:34:56", null, 1687747869032, 20230625, 235109], kind = INSERT}
     ]
   }
 }
@@ -46,7 +49,7 @@ transform {
   Sql {
     source_table_name = "fake"
     result_table_name = "fake1"
-    query = "select cast(id as STRING) as id, cast(id as INT) as id2, cast(id as DOUBLE) as id3 , cast(c1 as double) as c1_1, cast(c1 as DECIMAL(10,2)) as c1_2, cast(c2 as DATE) as c2_1, coalesce(c3,'Unknown') c3_1, ifnull(c3,'Unknown') c3_2, ifnull(nullif(name,'Joy Ding'),'NULL') name1, nullif(name,'Joy Ding_') name2 from fake"
+    query = "select cast(id as STRING) as id, cast(id as INT) as id2, cast(id as DOUBLE) as id3 , cast(c1 as double) as c1_1, cast(c1 as DECIMAL(10,2)) as c1_2, cast(c2 as DATE) as c2_1, coalesce(c3,'Unknown') c3_1, ifnull(c3,'Unknown') c3_2, ifnull(nullif(name,'Joy Ding'),'NULL') name1, nullif(name,'Joy Ding_') name2, cast(c4 as timestamp) as c4_1, cast(c4 as decimal(17,4)) as c4_2, cast(c5 as date) as c5, cast(c6 as time) as c6 from fake"
   }
 }
 
@@ -87,6 +90,13 @@ sink {
           ]
         },
         {
+          field_name = "c1_2"
+          field_type = "decimal(10,2)"
+          field_value = [
+            {equals_to = "12.40"}
+          ]
+        },
+        {
           field_name = "c2_1"
           field_type = "date"
           field_value = [
@@ -119,6 +129,34 @@ sink {
           field_type = "string"
           field_value = [
             {equals_to = "Joy Ding"}
+          ]
+        },
+        {
+          field_name = "c4_1"
+          field_type = "timestamp"
+          field_value = [
+            {equals_to = "2023-06-26T02:51:09.032"}
+          ]
+        },
+        {
+          field_name = "c4_2"
+          field_type = "decimal(17,4)"
+          field_value = [
+            {equals_to = "1687747869032.0000"}
+          ]
+        },
+        {
+          field_name = "c5"
+          field_type = "date"
+          field_value = [
+            {equals_to = "2023-06-25"}
+          ]
+        },
+        {
+          field_name = "c6"
+          field_type = "time"
+          field_value = [
+            {equals_to = "23:51:09"}
           ]
         }
       ]

--- a/seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/functions/SystemFunction.java
+++ b/seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/functions/SystemFunction.java
@@ -22,9 +22,11 @@ import org.apache.seatunnel.transform.exception.TransformException;
 
 import java.math.BigDecimal;
 import java.math.RoundingMode;
+import java.time.Instant;
 import java.time.LocalDate;
 import java.time.LocalDateTime;
 import java.time.LocalTime;
+import java.time.ZoneId;
 import java.util.List;
 
 public class SystemFunction {
@@ -96,6 +98,14 @@ public class SystemFunction {
                 if (v1 instanceof LocalTime) {
                     return LocalDateTime.of(LocalDate.now(), (LocalTime) v1);
                 }
+                if (v1 instanceof Long) {
+                    Instant instant = Instant.ofEpochMilli(((Long) v1).longValue());
+                    ZoneId zone = ZoneId.systemDefault();
+                    return LocalDateTime.ofInstant(instant, zone);
+                }
+                throw new TransformException(
+                        CommonErrorCode.UNSUPPORTED_OPERATION,
+                        String.format("Unsupported CAST AS type: %s", v2));
             case "DATE":
                 if (v1 instanceof LocalDateTime) {
                     return ((LocalDateTime) v1).toLocalDate();
@@ -103,16 +113,36 @@ public class SystemFunction {
                 if (v1 instanceof LocalDate) {
                     return v1;
                 }
+                if (v1 instanceof Integer) {
+                    int dateValue = ((Integer) v1).intValue();
+                    int year = dateValue / 10000;
+                    int month = (dateValue / 100) % 100;
+                    int day = dateValue % 100;
+                    return LocalDate.of(year, month, day);
+                }
+                throw new TransformException(
+                        CommonErrorCode.UNSUPPORTED_OPERATION,
+                        String.format("Unsupported CAST AS type: %s", v2));
             case "TIME":
                 if (v1 instanceof LocalDateTime) {
                     return ((LocalDateTime) v1).toLocalTime();
                 }
                 if (v1 instanceof LocalDate) {
-                    return LocalDateTime.of((LocalDate) v1, LocalTime.of(0, 0, 0));
+                    return LocalTime.of(0, 0, 0);
                 }
                 if (v1 instanceof LocalTime) {
-                    return LocalDateTime.of(LocalDate.now(), (LocalTime) v1);
+                    return v1;
                 }
+                if (v1 instanceof Integer) {
+                    int intTime = ((Integer) v1).intValue();
+                    int hour = intTime / 10000;
+                    int minute = (intTime / 100) % 100;
+                    int second = intTime % 100;
+                    return LocalTime.of(hour, minute, second);
+                }
+                throw new TransformException(
+                        CommonErrorCode.UNSUPPORTED_OPERATION,
+                        String.format("Unsupported CAST AS type: %s", v2));
             case "DECIMAL":
                 BigDecimal bigDecimal = new BigDecimal(v1.toString());
                 Integer scale = (Integer) args.get(3);

--- a/seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/functions/SystemFunction.java
+++ b/seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/functions/SystemFunction.java
@@ -104,7 +104,7 @@ public class SystemFunction {
                     return LocalDateTime.ofInstant(instant, zone);
                 }
                 throw new TransformException(
-                        CommonErrorCode.UNSUPPORTED_OPERATION,
+                        CommonErrorCodeDeprecated.UNSUPPORTED_OPERATION,
                         String.format("Unsupported CAST AS type: %s", v2));
             case "DATE":
                 if (v1 instanceof LocalDateTime) {
@@ -121,7 +121,7 @@ public class SystemFunction {
                     return LocalDate.of(year, month, day);
                 }
                 throw new TransformException(
-                        CommonErrorCode.UNSUPPORTED_OPERATION,
+                        CommonErrorCodeDeprecated.UNSUPPORTED_OPERATION,
                         String.format("Unsupported CAST AS type: %s", v2));
             case "TIME":
                 if (v1 instanceof LocalDateTime) {
@@ -141,7 +141,7 @@ public class SystemFunction {
                     return LocalTime.of(hour, minute, second);
                 }
                 throw new TransformException(
-                        CommonErrorCode.UNSUPPORTED_OPERATION,
+                        CommonErrorCodeDeprecated.UNSUPPORTED_OPERATION,
                         String.format("Unsupported CAST AS type: %s", v2));
             case "DECIMAL":
                 BigDecimal bigDecimal = new BigDecimal(v1.toString());

